### PR TITLE
[pipelining] Wire static eager pipeline execution

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.18.0) (oldest at bottom):
* #4543
* #4542
* #4541
* #4665
* #4623
* #4559
* #4652
* #4591
* #4592
* __->__ #4540
* #4539
* #4643

TorchTitan's eager pipeline runtime needs deterministic, precomputed contracts
before warmup: stage-boundary metadata, receive-buffer ownership, P2P topology,
and the maximum number of simultaneously unsharded local stages. Dynamic
metadata exchange or runtime buffer allocation prevents FakePG execution and
makes whole-step CUDA-graph capture difficult to reason about.

Wire the existing PyTorch pipelining facilities into TorchTitan:

- enable schedule-derived directed-edge P2P groups whenever pipeline
  parallelism is active;
- enable deterministic receive-buffer recycling for looped schedules;
- construct complete static decoder stage I/O metadata when the PP group is
  fake, failing clearly when a custom split cannot be represented;
- pass the configurable local-stage parameter-residency bound to the schedule;
- carry a fixed upper bound for varlen document metadata.

Directed P2P groups, receive-buffer reuse, and metadata selection are runtime
invariants, not user policy knobs. Their groups and assignments are finalized
before warmup; execution does not poll, recolor storage, or mutate the
schedule. The only public memory-policy knob added here is
`pipeline_parallel_max_param_unsharded_stages`.

This PR consumes a fake PP group when distributed initialization supplies one;
it does not define the fake or real-PP/fake-SPMD launch modes. Those topology
and launch contracts are introduced independently in #4652. Looped-schedule
CUDA-graph enablement is isolated in #4591.

Test Plan:
- `python -m pytest -q tests/unit_tests/cpu/test_no_new_cli_options.py tests/unit_tests/cpu/test_pipeline_parallel.py tests/unit_tests/cpu/test_config_manager.py tests/unit_tests/cpu/test_varlen_attention.py`
- Stack-level one-GPU pure-FakePG PP and four-GPU real-PP/fake-SPMD execution
  with repeated full CUDA-graph replay.

### CODE REVIEW

File: torchtitan/config/configs.py [hunk 1]
```diff
@@ -238,6 +239,18 @@ class ParallelismConfig:
...
+    """Place pipeline receives immediately before their first consumer."""
+
+    pipeline_parallel_reuse_recv_buffers: Annotated[bool, tyro.conf.Suppress] = False
+    """Reuse schedule-planned receive buffers across non-overlapping actions."""
+
+    pipeline_parallel_per_direction_p2p: Annotated[bool, tyro.conf.Suppress] = False
+    """Use one process group per directed physical-rank PP edge."""
+
+    pipeline_parallel_max_active_stages: Annotated[int, tyro.conf.Suppress] = 3
+    """Maximum FSDP stages kept unsharded by a looped pipeline schedule."""
```
Comment (tianyu-l) [inline torchtitan/config/configs.py]: this sounds legit, for others why would we make them configurable at all?

name is ambiguous, could be improved, e.g. `pipeline_parallel_max_param_unsharded_stages`

3 sounds too small as default? I feel default should be "everything being kept unsharded"

File: torchtitan/config/configs.py [hunk 2]
```diff
@@ -349,13 +364,16 @@ class CommConfig:
     save_traces_file_prefix: str = "rank_"
     """Flight recorder trace files prefix"""
 
-    mode: Literal["default", "fake_backend"] = "default"
+    mode: Literal["default", "fake_backend", "real_pp_fake_spmd_backend"] = "default"
```
Comment (tianyu-l) [inline torchtitan/config/configs.py]: "fake_backend" should cover both
- without PP it's purely fake
- with PP it's real PP fake spmd

File: torchtitan/distributed/parallel_dims.py [hunk 3]
```diff
@@ -80,11 +80,17 @@ class ParallelDims:
...
+    _pp_mesh_override: DeviceMesh | None = None
 
     @classmethod
     def from_config(
-        cls, parallelism_config: ParallelismConfig, world_size: int
+        cls,
+        parallelism_config: ParallelismConfig,
+        world_size: int,
+        *,
+        pp_mesh_override: DeviceMesh | None = None,
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py]: looks hacky, not clear why we need this to be passed in from outside

File: torchtitan/distributed/pipeline_parallel.py [hunk 4]
```diff
@@ -629,6 +641,7 @@ def _pipeline_module_split(
             device,
             group=pp_mesh.get_group("pp"),
             get_mesh=get_mesh,
+            pass_pipeline_metadata=pass_pipeline_metadata,
```
Comment (tianyu-l) [inline torchtitan/distributed/pipeline_parallel.py]: why introducing this flag as configurable? Shouldn't it be hardcoded to be `True`?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/config/configs.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/config/configs.py]: I also want fake_backend to have fake everything including PP, pp works with fake backend, these use case for real_pp_fake_spmd backed is for exercising pp comms locally.

File: torchtitan/config/configs.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/config/configs.py]: If the number of stages on a rank are too much keeping them all unsharded throughout is very costly for memory, in that case we can reshard the earlier stages, it's a memory configuration knob that is good to have given that we are moving towards larger models and gpu poor regimes

File: torchtitan/distributed/pipeline_parallel.py (hunk 4 above)
Comment (sanketpurandare) [reply torchtitan/distributed/pipeline_parallel.py]: This metadata is stage and microbatch id metadata that is only useful for Dist-MoE

File: torchtitan/distributed/parallel_dims.py (hunk 3 above)
Comment (sanketpurandare) [reply torchtitan/distributed/parallel_dims.py]: This is for passing the real pp mesh when using real_pp_fake_spmd backend, agree that it is hacky will do it properly

File: torchtitan/config/configs.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/config/configs.py]: I'm saying:
1. why we have other configs?
2. I agree with having this knob, but default 3 sounds arbitrary -- I wonder if we should default to unshard everything, and if users need, they can tune it.
3. rename this knob to `pipeline_parallel_max_param_unsharded_stages`

File: torchtitan/config/configs.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/config/configs.py]: I'd prefer we simplify things right now to piggyback `real_pp_fake_spmd_backend` onto `fake_backend`, as there's no other option yet. We can worry about that later.

File: torchtitan/distributed/pipeline_parallel.py (hunk 4 above)
Comment (tianyu-l) [reply torchtitan/distributed/pipeline_parallel.py]: Then let's introduce it in the DistMoE integration PR, which we need to discuss about

File: torchtitan/config/configs.py (hunk 1 above)
Comment (sanketpurandare) [reply torchtitan/config/configs.py]: Removed the other public pipeline-runtime policy knobs and retained only `pipeline_parallel_max_param_unsharded_stages`, using the suggested name. Its default keeps every local stage eligible to remain unsharded; an explicit positive limit is the memory-control override.

File: torchtitan/config/configs.py (hunk 2 above)
Comment (sanketpurandare) [reply torchtitan/config/configs.py]: Kept pure fake and real-PP/fake-SPMD as separate modes because they validate different contracts: pure fake models one rank of the complete logical topology without transport, while the hybrid mode runs real PP communication and fake SPMD collectives. The focused topology PR #4652 now returns one typed topology result with explicit validation rather than mutating a constructed mesh or falling back between modes. Its debugging guide documents the use cases, logical-rank formula, every launch variable, exact commands, represented memory, and limitations; the documented one-GPU pure-fake and two-GPU hybrid PP2/DP2 commands both completed one step.

File: torchtitan/distributed/parallel_dims.py (hunk 3 above)
Comment (sanketpurandare) [reply torchtitan/distributed/parallel_dims.py]: Removed the external PP-mesh override and post-construction mesh mutation. Distributed initialization now returns a typed topology that contains the real PP and fake-SPMD meshes, and `ParallelDims` consumes it at construction.

File: torchtitan/distributed/pipeline_parallel.py (hunk 4 above)
Comment (sanketpurandare) [reply torchtitan/distributed/pipeline_parallel.py]: Removed the public metadata flag. Pipeline stage and microbatch metadata is now capability-driven: setup marks only participating stages, the schedule supplies canonical kwargs, and a stage-root pre-hook consumes them before ordinary model forward.

File: torchtitan/models/common/attention.py [hunk 5]
```diff
@@ -144,6 +144,9 @@ class Config(InnerAttention.Config):
               - (W, 0): Sliding window causal - attend to at most W previous tokens.
         """
 
+        max_num_documents: int | None = None
+        """Upper bound on packed documents used for fixed-shape metadata."""
+
```
Comment (tianyu-l) [inline torchtitan/models/common/attention.py:149]: accidental?

File: torchtitan/distributed/pipeline_parallel.py [hunk 6]
```diff
@@ -572,13 +609,108 @@ def _get_pp_rank_to_stage_indices_mapping(
...
+def _build_decoder_stage_io(
+    *,
+    parallel_dims: ParallelDims,
+    parallelism: ParallelismConfig,
+    training: TrainingConfig,
+    model_config: Decoder.Config,
+    lm_head_in_loss: bool,
+) -> _DecoderStageIO:
+    """Build static metadata for tensors crossing decoder stage boundaries."""
+    num_tokens, cp_remainder = divmod(
```
Comment (tianyu-l) [inline torchtitan/distributed/pipeline_parallel.py:644]: https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L358

here we require another factor of 2 because of head-tail load balancing splitting

File: torchtitan/distributed/pipeline_parallel.py [hunk 7]
```diff
@@ -572,13 +609,108 @@ def _get_pp_rank_to_stage_indices_mapping(
         raise ValueError(f"Unknown style {style}")
 
 
+@dataclasses.dataclass(frozen=True)
+class _DecoderStageIO:
+    """Example tensors describing decoder pipeline boundaries."""
+
+    root_input: torch.Tensor
```
Comment (tianyu-l) [inline torchtitan/distributed/pipeline_parallel.py:616]:
```suggestion
    decoder_input: torch.Tensor
```

File: torchtitan/distributed/pipeline_parallel.py [hunk 8]
```diff
@@ -572,13 +609,108 @@ def _get_pp_rank_to_stage_indices_mapping(
...
+    hidden: torch.Tensor
+    final_output: torch.Tensor
```
Comment (tianyu-l) [inline torchtitan/distributed/pipeline_parallel.py:618]:
```suggestion
    decoder_output: torch.Tensor
```
